cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,1258 @@
+  static constexpr auto MANT_DIG  = std::numeric_limits<ftype>::digits;
+
+public:
+  /// Binned floating-point maximum index
+  enum
+  {
+    MAXINDEX = ((MAX_EXP - MIN_EXP + MANT_DIG - 1) / BIN_WIDTH) - 1
+  };
+
+  // The maximum floating-point fold supported by the library
+  static constexpr auto MAXFOLD = MAXINDEX + 1;
+
+private:
+  /// Binned floating-point compression factor
+  /// This factor is used to scale down inputs before deposition into the bin of
+  /// highest index
+  static constexpr auto COMPRESSION = 1.0 / (1 << (MANT_DIG - BIN_WIDTH + 1));
+  /// Binned double precision expansion factor
+  /// This factor is used to scale up inputs after deposition into the bin of
+  /// highest index
+  static constexpr auto EXPANSION = 1.0 * (1 << (MANT_DIG - BIN_WIDTH + 1));
+  static constexpr auto EXP_BIAS  = MAX_EXP - 2;
+  static constexpr auto EPSILON   = std::numeric_limits<ftype>::epsilon();
+  /// Binned floating-point deposit endurance
+  /// The number of deposits that can be performed before a renorm is necessary.
+  /// Applies also to binned complex double precision.
+  static constexpr auto ENDURANCE = 1 << (MANT_DIG - BIN_WIDTH - 2);
+
+  /// Return a binned floating-point bin
+  __host__ __device__ inline ftype binned_bins(int index) const
+  {
+#ifdef __CUDA_ARCH__
+    ftype* bins = get_shared_bin_array<ftype, MAXINDEX + MAXFOLD>();
+    if (index >= MAXINDEX + MAXFOLD)
+    {
+      return static_cast<ftype>(bins[MAXINDEX + MAXFOLD - 1]);
+    }
+    if (index < 0)
+    {
+      return static_cast<ftype>(bins[0]);
+    }
+    return static_cast<ftype>(bins[index]);
+#else
+    static float bins_host[MAXINDEX + MAXFOLD] = {};
+
+    for (int i = 0; i < MAXINDEX + MAXFOLD; ++i)
+    {
+      bins_host[i] = RFA_bins<ftype>::initialize_bins(i);
+    }
+
+    if (index >= MAXINDEX + MAXFOLD)
+    {
+      return static_cast<ftype>(bins_host[MAXINDEX + MAXFOLD - 1]);
+    }
+    if (index < 0)
+    {
+      return static_cast<ftype>(bins_host[0]);
+    }
+    return static_cast<ftype>(bins_host[index]);
+#endif
+  }
+
+  /// Get the bit representation of a float
+  __host__ __device__ static inline uint32_t& get_bits(float& x)
+  {
+    return *reinterpret_cast<uint32_t*>(&x);
+  }
+  /// Get the bit representation of a double
+  __host__ __device__ static inline uint64_t& get_bits(double& x)
+  {
+    return *reinterpret_cast<uint64_t*>(&x);
+  }
+  /// Get the bit representation of a const float
+  __host__ __device__ static inline uint32_t get_bits(const float& x)
+  {
+    return *reinterpret_cast<const uint32_t*>(&x);
+  }
+  /// Get the bit representation of a const double
+  __host__ __device__ static inline uint64_t get_bits(const double& x)
+  {
+    return *reinterpret_cast<const uint64_t*>(&x);
+  }
+
+  /// Return primary vector value const ref
+  __host__ __device__ __forceinline__ const ftype& primary(int i) const
+  {
+    if constexpr (FOLD <= MAX_JUMP)
```

Comment:
Question: is the use of C++17 blessed? IIUC, we can require C++17 for entirely new features in CUB/Thrust, but everything that changes existing infrastructure has to remain compatible with C++11 until *the big drop*.

Reply:
For new API, it's a compromise. If it takes little effort, we support C++11. It it'll take us a month to backport new API to C++11, we can avoid doing that.

Reply:
I have changed significant code to compare the binary and determine if this is taking a toll on performance. Although I am guessing it should not :-)

Reply:
This is not a question of performance, it's rather about your tolerance of pain :) (and size of user base that can immediately use this feature).